### ISSUE TITLE
`cargo generate` changes project names containing `_`

### ISSUE THREAD

**Issue by mohammadfawaz** [FuelLabs/sway#3050]
It looks like when we run `cargo generate --init fuellabs/sway templates/sway-test-rs --name fuel_project`, the generated test harness has `-` instead of the underscore in the name, and so the `abigen!` macro looks like `abigen!(MyContract, "out/debug/fuel-project-abi.json");` The project name in `Cargo.toml` is also exhibiting the same behavior. The reason for this is that we use the Cargo generate builtin placeholder`{{project-name}}` for the project name. If we wanted underscore, we would have to use the placeholder `{{crate_name}}` but that would change `-` to `_`, so we'd have a similar problem.

I don't see any default placeholders that would preserve the name 100% so we may have to use a custom placeholder. 

It would also be nice if the user does not have to specify the `name` field at all since the name _has_ to match the name of the current directory. If cargo generate can just set that name automatically, that would be great. I think placeholder are allowed to have defaults.

**Comment by mohammadfawaz** [comment]
@mitchmindtree not sure if you had a chance to look at this. What are you thoughts?

**Comment by mitchmindtree** [comment]
Hey thanks for opening this @mohammadfawaz!

> It looks like when we run cargo generate --init fuellabs/sway templates/sway-test-rs --name fuel_project, the generated test harness has - instead of the underscore in the name, and so the abigen! macro looks like abigen!(MyContract, "out/debug/fuel-project-abi.json");

Hmmm this sounds a bit like a `cargo generate` bug? It could be worth checking that this is intentional first, especially as it's valid for rust project's to contain `_`s.

> It would also be nice if the user does not have to specify the name field at all since the name has to match the name of the current directory.

I don't believe a project's directory has to have the same name as the project itself, but I agree it would be much nicer to do something like this! Perhaps retrieving the name from the `name` field within the current directory's `Forc.toml` instead.

**Comment by mitchmindtree** [comment]
> > I don't believe a project's directory has to have the same name as the project itself
>
> So in principle it shouldn't, but here's what I currently get

Yes I understand the test path issue. I only meant to clarify that we should not use the current directory name when addressing this as doing so would be unreliable (i.e. Sway has not rule that the current directory name _must_ match the project name), so we should retrieve the name from the `Forc.toml` instead.

**Comment by mohammadfawaz** [comment]
> so we should retrieve the name from the `Forc.toml` instead.

Ah, yes this makes sense to me now :)

**Comment by bingcicle** [comment]
It seems like you can specify `--force` to enforce that the project name doesn't auto convert to kebab-case. From output of `cargo generate --help`:

```console
$ cargo generate --help
...
    -n, --name <NAME>
            Directory to create / project name; if the name isn't in kebab-case, it will be
            converted to kebab-case unless `--force` is given
...
```

By adding this flag to the instructions on the [Testing with Rust](https://fuellabs.github.io/sway/v0.26.0/testing/testing-with-rust.html#3-generate-the-test-harness) page I'm able to execute `forc build` and `cargo test` as described in the docs.

If this is a tenable solution, perhaps we just need to update the docs with this flag, and describe what `--force` does?

**Comment by mohammadfawaz** [comment]
Should we wrap everything up in a new simple `forc` command that generates a default test harness with the default name?

**Comment by bingcicle** [comment]
> Should we wrap everything up in a new simple forc command that generates a default test harness with the default name?

Yeah this sounds good, something like `forc generate-test` or `forc generate-harness`?

Then we can also add in logic that reads the name from `Forc.toml` and use that as an input to `cargo-generate`.

**Comment by mitchmindtree** [comment]
@mohammadfawaz @bingcicle I don't think we should wrap this up with a forc command (at least not a native one in `forc` itself). The whole point of removing the Rust integration testing from the native `forc test` command was to remove the implicit dependency on the Rust toolchain. If we're to add a subcommand like this, then not *only* do we depend on `cargo` during `forc`'s runtime again, but now we'd also depend on the `cargo generate` command. Rust integration testing is only relevant to developers using Sway and Rust together, but not relevant to folks using Sway alongside any other language (e.g. Typescript, Go, whatever the new hotness is).

> Then we can also add in logic that reads the name from Forc.toml and use that as an input to cargo-generate.

I think `cargo-generate` has some kind of scripting support for templates - perhaps we could investigate this as an option instead?

**Comment by mitchmindtree** [comment]
> Perhaps for now we should add --force as a flag + a short explanation to our cargo generate command within our docs anyway while we figure out a more permanent/better solution?

Yeah this sounds like a good approach for now.

Perhaps we can also open an issue or contribute a PR upstream at the `cargo-generate` repo for optionally inheriting the name from the current project in the case that `--init` was specified? I imagine this would be a common request for anyone creating `cargo generate` templates that are designed to be used with the `--init` flag. I'm kind of surprised there's no solution for this upstream already!

**Comment by bingcicle** [comment]
I'll open an issue describing our use case for that upstream :+1:

**Comment by bingcicle** [comment]
I will mark this as closed with #3110, but have opened a new issue at #3111 with a summary of the discussion here.

**Comment by mohammadfawaz** [comment]
I like the idea of `cargo generate --init`.
